pdl_syntax.md

```diff
@@ -0,0 +1,828 @@
+  booleans: map[string, boolean]
+  strings: map[string, string]
+  bytes: map[string, bytes]
+}
+```
+Custom Types Map :
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+import com.linkedin.pegasus.generator.test.idl.customtypes.CustomInt
+
+record WithCustomTypesMap {
+  ints: map[string, CustomInt]
+}
+```
+
+Complex Types Map :
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+import com.linkedin.pegasus.generator.test.idl.enums.Fruits
+import com.linkedin.pegasus.generator.test.idl.records.Empty
+import com.linkedin.pegasus.generator.test.idl.records.Simple
+import com.linkedin.pegasus.generator.test.idl.`fixed`.Fixed8
+
+record WithComplexTypesMap {
+  empties: map[string, Empty]
+  fruits: map[string, Fruits]
+  arrays: map[string, array[Simple]]
+  maps: map[string, map[string, Simple]]
+  unions: map[
+    string,
+    typeref WithComplexTypesMapUnion = union[int, string, Simple]
+  ]
+  `fixed`: map[string, Fixed8]
+}
+```
+
+**Note**: 
+The key must always be "string".
+
+
+## Union Type
+
+A union type may be defined with any number of member types. Each member may be any pegasus type except union type. 
+Each member can be primitive, record, enum map or array type.
+
+For example :
+```
+namespace com.linkedin.pegasus.generator.test.idl.unions
+
+record WithPrimitivesUnion {
+  `union`: union[int, long, float, double, boolean, string, bytes]
+}
+```
+
+The member type names also serve as the “member keys” (also called as “union tags”), and identify which union member type data holds.
+To define a field of a record containing a union of two other records, we would define:
+```
+namespace com.linkedin.pegasus.generator.examples
+
+record Question {
+  answerFormat: union[MultipleChoice, TextEntry]
+}
+```
+
+### Union with aliases
+Union members can optionally be given an alias. Aliases can be used to create unions with members of the same type or to give better naming for union members.
+
+Aliased unions would be defines as :
+```
+union [alias: type, ...]
+```
+
+For example :
+```
+namespace com.linkedin.pegasus.generator.examples
+
+record Question {
+ answerFormat: union[   
+   multipleChoice: MultipleChoice,
+   shortAnswer: string,
+   longAnswer: string
+ ]
+}
+```
+In the above example, the union answerFormat has three members, with two string type members differentiated using the aliases (shortAnswer and longAnswer). When aliases are used, the alias becomes the "member key" for the union members and will be used in the wire format. 
+
+## Enum Type
+
+Enums types may contain any number of symbols. 
+
+For example :
+
+```
+namespace com.linkedin.pegasus.generator.examples
+
+enum Fruits {
+  APPLE
+  BANANA
+  ORANGE
+  PINEAPPLE
+}
+```
+
+Enums can be referenced in other schemas by name.
+
+For example :
+
+```
+namespace com.linkedin.pegasus.generator.examples
+
+record FruitBasket {
+  fruit: Fruits
+}
+```
+
+Enums can also be referenced in other schemas by inlining their type definition. 
```

Comment:
This is not referencing, this is actually inline defining of the enum.

Reply:
Updated.